### PR TITLE
Caching layer

### LINKED ISSUE #5427: [Data Object] Relation Types DB Caching Layer

### PR DESCRIPTION
**Pull request by weisswurstkanone**
resolves #5427

### CODE REVIEW

**Comment by dpfaffenbauer** [comment]
where does the cache get invalidated?

**Comment by brusch** [comment]
When the object gets updated (on save). The raw data is directly on the object as a proeprty, so no additional invalidation is needed.

File: models/DataObject/ClassDefinition.php [hunk 1]
```diff
@@ -1283,4 +1299,24 @@ public function getLinkGenerator()
...
+    /**
+     * @return bool
+     */
+    public function getCacheRawRelationData(): bool
+    {
+        return $this->cacheRawRelationData;
+    }
+
+    /**
+     * @param $cacheRawRelationData
```
Comment (blankse) [inline models/DataObject/ClassDefinition.php]: `@param bool $cacheRawRelationData`

File: models/DataObject/ClassDefinition/Service.php [hunk 2]
```diff
@@ -242,9 +242,8 @@ public static function importObjectBrickFromJson($objectBrick, $json, $throwExce
     /**
      * @param array $array
      * @param bool $throwException
-     * @param bool $insideLocalizedField
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Service.php]: This param exists

File: models/DataObject/ClassDefinition/Service.php [hunk 3]
```diff
@@ -242,9 +242,8 @@ public static function importObjectBrickFromJson($objectBrick, $json, $throwExce
...
      *
-     * @return mixed
+     * @return bool
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Service.php]: Can also be `DataObject\ClassDefinition\Data`

File: models/DataObject/Concrete.php [hunk 4]
```diff
@@ -881,4 +881,59 @@ public static function getLatestVersionByObjectIdAndLatestModificationDate($obje
 
         return;
     }
+
+    /**
+     * @param $descriptor
```
Comment (blankse) [inline models/DataObject/Concrete.php]: `@param array $descriptor`

**Comment by dpfaffenbauer** [comment]
what is the actual idea for this? I am obviously missing the point....

**Comment by markus-moser** [comment]
@weisswurstkanone what happens when an object get saved? does it load the relations data always "fresh" from the database or is the cached used for it?

**Comment by weisswurstkanone** [comment]
@blankse , thx!

File: models/DataObject/Concrete.php [hunk 5]
```diff
@@ -883,7 +883,7 @@ public static function getLatestVersionByObjectIdAndLatestModificationDate($obje
     }
 
     /**
-     * @param $descriptor
+     * @param array $descriptor
      * @return array|mixed|mixed[]
```
Comment (NiklasBr) [inline models/DataObject/Concrete.php]: This method will now return these two types:
`array`: line 921
`mixed[]`: 936

File: models/DataObject/Concrete.php [hunk 6]
```diff
@@ -881,4 +881,62 @@ public static function getLatestVersionByObjectIdAndLatestModificationDate($obje
...
+            foreach ($descriptor as $column => $expectedValue) {
+                if (is_string($expectedValue)) {
+                    $trimmed = rtrim($expectedValue, '%');
+                    if (strlen($trimmed) < strlen($expectedValue)) {
+                        $likes[$column] = $trimmed;
+                    }
+                }
+            }
+
+            $filterFn = function($row) use ($descriptor, $likes) {
```
Comment (brusch) [inline models/DataObject/Concrete.php]: @weisswurstkanone 
for a micro-optimzation we can use `static function` here as `$this` context is not needed

**Comment by brusch** [comment]
@dpfaffenbauer  see docs page: 

> Enable this if you want to retrieve all rows of the affected `object_relation_` table in one go and cache the results.
This can be beneficial if your model contains many lazy loaded fields that are frequently used.

😊 

Imho this should be an advantage in any case, however we'd like to get some more real life experiences on that fist before we make ir the default behavior.

**Pull request merged**